### PR TITLE
bpf: Remove duplicate code in bpf_sock.c

### PR DESCRIPTION
**Pull request by brb**
This commit refactors `__sock{4,6}_xlate()` functions in a way that they can be used to replace `__sock{4,6}_snd_xlate()`.

The refactoring also allows us to get rid of the `sock6_xlate_snd_v4_in_v6()` function.

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

**Review by tklauser — CHANGES_REQUESTED** [review]
Would it make sense to add `-DHOOK_CONNECT` to `LB_OPTIONS` in `bpf/Makefile` so it gets compile-tested too in various option combinations?

**Review by borkmann — CHANGES_REQUESTED** [review]
Makes sense, but why do we need even more ifdef wrt `HOOK_CONNECT` at all? Either we could leave the sanity check in place (which doesn't hurt much either) or if we really want to remove it then via `const bool` so compiler optimizes it out. I'd probably do the former so `__sock4_xlate()` could stay unchanged, for example.

**Comment by brb** [comment]
> Makes sense, but why do we need even more ifdef wrt HOOK_CONNECT at all

@borkmann It was just to save a few CPU cycles considering that the unnecessary checks would be in the fastpath in the case of UDP. If you think that's not worth doing, then I can remove the ifdefs from the checks.

**Comment by borkmann** [comment]
> > Makes sense, but why do we need even more ifdef wrt HOOK_CONNECT at all
> 
> @borkmann It was just to save a few CPU cycles considering that the unnecessary checks would be in the fastpath in the case of UDP. If you think that's not worth doing, then I can remove the ifdefs from the checks.

But in such case I'd just go for `const bool udp_only` or such, so we have less ifdef in the code and compiler can optimize it out.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
@tklauser `HOOK_CONNECT` has been removed.

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -286,14 +287,19 @@ static __always_inline int __sock4_xlate(struct bpf_sock_addr *ctx,
...
+	}
+
+	if (sock4_update_revnat(ctx_full, backend, &key,
 				slave_svc) < 0) {
 		update_metrics(0, METRIC_EGRESS, REASON_LB_REVNAT_UPDATE);
 		return -ENOMEM;
 	}
 
-	ctx->user_ip4	= backend->address;
+update_dst: __maybe_unused
```
Comment (tklauser) [inline bpf/bpf_sock.c]: Why is it needed to mark this as `__maybe_unused`? It seems it's always used.

File: bpf/bpf_sock.c [hunk 2]
```diff
@@ -716,123 +673,38 @@ static __always_inline int __sock6_xlate(struct bpf_sock_addr *ctx)
 		return -ENOENT;
 	}
 
-	if (ctx->protocol != IPPROTO_TCP &&
-	    sock6_update_revnat(ctx, backend, &key,
+	if (!udp_only && ctx->protocol == IPPROTO_TCP) {
+		goto update_dst;
```
Comment (tklauser) [inline bpf/bpf_sock.c:677]: Same comment as for `__sock4_xlate` applies here regarding `goto` and `update_dst` jump label with `__maybe_unused`.

**Review by tklauser — APPROVED** [review]
Even nicer without the `HOOK_CONNECT` define! Two small nits/questions inline. Otherwise LGTM.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Review by borkmann — APPROVED** [review]

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
CI hit the flake #10669. As my code changes are tested by the flaking test, re-running the CI.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Pull request merged**